utils/src/Vulkan/Utils/Reflect.hs

```diff
@@ -0,0 +1,602 @@
+  , vertexAttributeDescriptions
+  }
+ where
+  vertexAttributes =
+    join . V.map makeVertexAttribute $ inputs :: Vector VertexAttribute
+  vertexAttributeDescriptions =
+    makeVertexInputAttributeDescriptions vertexAttributes :: Vector
+        VertexInputAttributeDescription
+  vertexBindingDescriptions =
+    makeVertexInputBindingDescriptions vertexAttributes :: Vector
+        VertexInputBindingDescription
+
+makeVertexInputBindingDescriptions
+  :: Vector VertexAttribute -> Vector VertexInputBindingDescription
+makeVertexInputBindingDescriptions =
+  V.fromList
+    . map makeVertexInputBindingDescription
+    . calculate
+    . groupBy ((==) `on` fst)
+    . sortOn fst
+    . extract
+ where
+  extract :: Vector VertexAttribute -> [("binding" ::: Int, "size" ::: Int)]
+  extract =
+    map
+        (   (,)
+        .   fromIntegral
+        .   (binding :: VertexAttribute -> Word32)
+        <*> fromIntegral
+        .   size
+        )
+      . V.toList
+  calculate
+    :: [[("binding" ::: Int, "size" ::: Int)]]
+    -> [("binding" ::: Int, "stride" ::: Int)]
+  calculate = map (liftA2 (,) (fst . head) (sum . (snd <$>)))
+
+makeVertexInputBindingDescription
+  :: ("binding" ::: Int, "stride" ::: Int) -> VertexInputBindingDescription
+makeVertexInputBindingDescription (binding, stride) = zero
+  { binding   = fromIntegral binding
+  , stride    = fromIntegral stride
+  , inputRate = VERTEX_INPUT_RATE_VERTEX
+  }
+
+makeVertexAttribute :: Input -> Vector VertexAttribute
+makeVertexAttribute Input {..} = V.map
+  (\i -> VertexAttribute { binding  = 0
+                         , location = fromIntegral . (+ location) $ i
+                         , size
+                         , format
+                         }
+  )
+  [0 .. count - 1]
+ where
+  count          = maybe 1 V.sum array :: Int
+  (size, format) = convertVertexAttributeType type'
+
+makeVertexInputAttributeDescriptions
+  :: Vector VertexAttribute -> Vector VertexInputAttributeDescription
+makeVertexInputAttributeDescriptions =
+  V.fromList
+    . join
+    . map process
+    . groupBy ((==) `on` (binding :: VertexAttribute -> Word32))
+    . V.toList
+ where
+  process :: [VertexAttribute] -> [VertexInputAttributeDescription]
+  process = snd . foldl
+    (\(nextOffset, acc) cur -> liftA2
+      (,)
+      fst
+      ((acc ++) . pure . snd)
+      (makeVertexInputAttributeDescription nextOffset cur)
+    )
+    (0, [])
+
+makeVertexInputAttributeDescription
+  :: Word32 -> VertexAttribute -> (Word32, VertexInputAttributeDescription)
+makeVertexInputAttributeDescription offset VertexAttribute {..} =
+  (offset + size, zero { binding, location, offset, format })
+
+reflect
+  :: MonadIO m
+  => ShaderStage
+  -> "code" ::: Text
+  -> m (B.ByteString, BL.ByteString)
+reflect shaderStage code =
+  liftIO . withSystemTempDirectory "th-spirv" $ \dir -> do
+    let stage  = T.unpack . shaderStageToText $ shaderStage
+    let shader = dir </> "glsl." <> stage
```

Comment:
There's `<.>` operator for adding extensions.

Shouldn't this be `stage <.> "glsl"`, like `spv` below?